subsys/usb/device_next/class/bt_hci.c

```diff
@@ -0,0 +1,507 @@
+	if (!buf) {
+		return NULL;
+	}
+
+	bi = udc_get_buf_info(buf);
+	memset(bi, 0, sizeof(struct udc_buf_info));
+	bi->ep = ep;
+
+	return buf;
+}
+
+static void bt_hci_tx_sync_in(struct usbd_class_node *const c_nd,
+			      struct net_buf *const bt_buf, const uint8_t ep)
+{
+	struct bt_hci_data *hci_data = c_nd->data->priv;
+	struct net_buf *buf;
+
+	buf = bt_hci_buf_alloc(ep);
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate buffer");
+		return;
+	}
+
+	net_buf_add_mem(buf, bt_buf->data, bt_buf->len);
+	usbd_ep_enqueue(c_nd, buf);
+	k_sem_take(&hci_data->sync_sem, K_FOREVER);
+	net_buf_unref(buf);
+}
+
+static void bt_hci_tx_thread(void *p1, void *p2, void *p3)
+{
+	struct usbd_class_node *const c_nd = p1;
+
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	while (true) {
+		struct net_buf *bt_buf;
+		uint8_t ep;
+
+		bt_buf = net_buf_get(&bt_hci_tx_queue, K_FOREVER);
+
+		switch (bt_buf_get_type(bt_buf)) {
+		case BT_BUF_EVT:
+			ep = bt_hci_get_int_in(c_nd);
+			break;
+		case BT_BUF_ACL_IN:
+			ep = bt_hci_get_bulk_in(c_nd);
+			break;
+		default:
+			LOG_ERR("Unknown type %u", bt_buf_get_type(bt_buf));
+			continue;
+		}
+
+
+		bt_hci_tx_sync_in(c_nd, bt_buf, ep);
+		net_buf_unref(bt_buf);
+	}
+}
+
+static void bt_hci_rx_thread(void *a, void *b, void *c)
+{
+	while (true) {
+		struct net_buf *buf;
+		int err;
+
+		/* FIXME: Do we need a separate thread for bt_send()? */
+		buf = net_buf_get(&bt_hci_rx_queue, K_FOREVER);
+
+		err = bt_send(buf);
+		if (err) {
+			LOG_ERR("Error sending to driver");
+			net_buf_unref(buf);
+		}
+	}
+}
+
+static int bt_hci_acl_out_start(struct usbd_class_node *const c_nd)
+{
+	struct bt_hci_data *hci_data = c_nd->data->priv;
+	struct net_buf *buf;
+	uint8_t ep;
+	int ret;
+
+	if (!atomic_test_bit(&hci_data->state, BT_HCI_CLASS_ENABLED)) {
+		return -EPERM;
+	}
+
+	if (atomic_test_and_set_bit(&hci_data->state, BT_HCI_ACL_RX_ENGAGED)) {
+		return -EBUSY;
+	}
+
+	ep = bt_hci_get_bulk_out(c_nd);
+	buf = bt_hci_buf_alloc(ep);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	ret = usbd_ep_enqueue(c_nd, buf);
+	if (ret) {
+		LOG_ERR("Failed to enqueue net_buf for 0x%02x", ep);
+		net_buf_unref(buf);
+	}
+
+	return  ret;
+}
+
+static uint16_t hci_pkt_get_len(struct net_buf *const buf,
+				const uint8_t *data, const size_t size)
+{
+	size_t hdr_len = 0;
+	uint16_t len = 0;
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_CMD: {
+		struct bt_hci_cmd_hdr *cmd_hdr;
+
+		hdr_len = sizeof(*cmd_hdr);
+		cmd_hdr = (struct bt_hci_cmd_hdr *)data;
+		len = cmd_hdr->param_len + hdr_len;
+		break;
+	}
+	case BT_BUF_ACL_OUT: {
+		struct bt_hci_acl_hdr *acl_hdr;
+
+		hdr_len = sizeof(*acl_hdr);
+		acl_hdr = (struct bt_hci_acl_hdr *)data;
+		len = sys_le16_to_cpu(acl_hdr->len) + hdr_len;
+		break;
+	}
+	case BT_BUF_ISO_OUT: {
+		struct bt_hci_iso_hdr *iso_hdr;
+
+		hdr_len = sizeof(*iso_hdr);
+		iso_hdr = (struct bt_hci_iso_hdr *)data;
+		len = bt_iso_hdr_len(sys_le16_to_cpu(iso_hdr->len)) + hdr_len;
+		break;
+	}
```

Comment:
Given that this driver otherwise doesn't support ISO, should this be removed?

Reply:
Maybe, but then it must later be added for H4 support.

Reply:
Seems odd to only partially add ISO/H4 support